internal/jobcontainers/jobcontainer.go

```diff
@@ -0,0 +1,449 @@
+	}
+
+	defer func() {
+		if err != nil {
+			process.Close()
+		}
+	}()
+
+	if err = process.Start(); err != nil {
+		return nil, errors.Wrap(err, "failed to start host process")
+	}
+
+	if err = c.job.Assign(uint32(process.Pid())); err != nil {
+		return nil, errors.Wrap(err, "failed to assign process to job object")
+	}
+
+	// Assign the first process made as the init process of the container.
+	c.init.initDoOnce.Do(func() {
+		c.init.proc = process
+		close(c.init.initBlock)
+	})
+
+	// Wait for process exit
+	go c.pollJobMsgs(ctx)
+	go process.waitBackground(ctx)
+	return process, nil
+}
+
+func (c *JobContainer) Modify(ctx context.Context, config interface{}) (err error) {
+	return errors.New("modify not supported for job containers")
+}
+
+// Release unmounts all of the container layers. Safe to call multiple times, if no storage
+// is mounted this call will just return nil.
+func (c *JobContainer) Release(ctx context.Context) error {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	log.G(ctx).WithFields(logrus.Fields{
+		"id":   c.id,
+		"path": c.sandboxMount,
+	}).Warn("removing sandbox volume mount")
+
+	if c.sandboxMount != "" {
+		if err := removeSandboxMountPoint(ctx, c.sandboxMount); err != nil {
+			return errors.Wrap(err, "failed to remove sandbox volume mount path")
+		}
+		if err := layers.UnmountContainerLayers(ctx, c.spec.Windows.LayerFolders, "", nil, layers.UnmountOperationAll); err != nil {
+			return errors.Wrap(err, "failed to unmount container layers")
+		}
+		c.sandboxMount = ""
+	}
+	return nil
+}
+
+// Start starts the container. There's nothing to "start" for job containers, so this just
+// sets the start timestamp.
+func (c *JobContainer) Start(ctx context.Context) error {
+	c.startTimestamp = time.Now()
+	return nil
+}
+
+// Close closes any open handles.
+func (c *JobContainer) Close() error {
+	if err := c.job.Close(); err != nil {
+		return err
+	}
+	c.closedWaitOnce.Do(func() {
+		c.waitError = hcs.ErrAlreadyClosed
+		close(c.waitBlock)
+	})
+	return nil
+}
+
+// ID returns the ID of the container. This is the name used to create the job object.
+func (c *JobContainer) ID() string {
+	return c.id
+}
+
+// Shutdown gracefully shuts down the container.
+func (c *JobContainer) Shutdown(ctx context.Context) error {
+	log.G(ctx).WithField("id", c.id).Debug("shutting down job container")
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*5)
+	defer cancel()
+	return c.shutdown(ctx)
+}
+
+// shutdown will loop through all the pids in the container and send a signal to exit.
+// If there are no processes in the container it will early return nil.
+// If the all processes exited message is not received within the context timeout set, it will
+// terminate the job.
+func (c *JobContainer) shutdown(ctx context.Context) error {
+	pids, err := c.job.Pids()
+	if err != nil {
+		return errors.Wrap(err, "failed to get pids in container")
+	}
+
+	if len(pids) == 0 {
+		return nil
+	}
+
+	for _, pid := range pids {
```

Comment:
Are you sure this matches the behavior of other container types? My understanding was in general we should just signal the init process.

Reply:
For regular Windows containers starting from `HcsShutdownComputeSystem` we end up all the way in cexecsvc which ultimately calls https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-initiatesystemshutdownexa at the end of the call chain. This is the closest I could find to that logic without turning off the host machine 😆